app/internal/statistics.py

```diff
@@ -0,0 +1,481 @@
+            (func.avg(event_duration) * NIN_IN_DAY),
+        )
+        .join(user_to_event)
+        .filter(by_user_id)
+        .filter(get_date_filter_between_dates(params[2], params[3]))
+        .all()
+    )
+    if events_duration_statistics[0][0]:
+        return EventsDurationStatistics(
+            shortest_event=round(events_duration_statistics[0][0]),
+            longest_event=round(events_duration_statistics[0][1]),
+            average_event=round(events_duration_statistics[0][2]),
+        )
+    return EventsDurationStatistics(
+        shortest_event=0, longest_event=0, average_event=0
+    )
+
+
+def get_events_duration_statistics(*params) -> Dict[str, Dict[str, float]]:
+    """calculate statistics for events durations relevant for
+        the requested user and the requested date range.
+        all logic is performed in the db, while the rest is in the code.
+
+    Args: list of parameters:
+        params[0] - db: db session.
+        params[1] - userid: requested user id number for statistics.
+        params[2] - start: start of date range.
+        params[3] - end: end of date range.
+
+    Returns: json with
+        events_duration_statistics: (for requested date range, in minutes).
+            shortest_event: shortest event the user has.
+            longest_event: longest event the user has.
+            average_event: average event the user has.
+    """
+    events_duration_statistics = get_events_duration_statistics_from_db(
+        *params
+    )
+    return {
+        "events_duration_statistics": {
+            "shortest_event": events_duration_statistics.shortest_event,
+            "longest_event": events_duration_statistics.longest_event,
+            "average_event": events_duration_statistics.average_event,
+        }
+    }
+
+
+def get_participants_statistics(
+    *params,
+) -> Dict[str, Dict[str, Union[str, int]]]:
+    """calculate statistics for events participants relevant for
+        the requested user and the requested date range.
+        part of the logic is performed in the db,
+        while the rest is in the code.
+
+    Args: list of parameters:
+        params[0] - db: db session.
+        params[1] - userid: requested user id number for statistics.
+        params[2] - start: start of date range.
+        params[3] - end: end of date range.
+
+    Returns: json with
+        max_events: maximum number of events the user has
+            with same participant.
+        participant_name: relevant participant name.
+    """
+    by_user_id = UserEvent.user_id == params[1]
+    by_not_user_id = UserEvent.user_id != params[1]
+    user_to_event = (UserEvent, UserEvent.event_id == Event.id)
+    participant_count = func.count(UserEvent.user_id)
+    subquery = (
+        params[0]
+        .query(Event.id)
+        .join(user_to_event)
+        .filter(by_user_id)
+        .filter(get_date_filter_between_dates(params[2], params[3]))
+        .subquery()
```

Comment:
Is this needed?

Reply:
I think that since we use a query on a result of another query:
The sub query gets all relevant events for the user (owner or participant), while the outer query gets participants for those events.
Therefore, in  my opinion, this syntax suits this need.